### DISCUSSION TITLE
Cannot override shortcuts with insert key

### DISCUSSION THREAD

**Discussion by Tim-Cao** [Textualize/textual#4050] [Q&A] [answered]
```python
from textual import on, work
from textual.app import App, ComposeResult, events
from textual.widgets import *
from textual.binding import Binding

class ScriptApp(App):
    BINDINGS = [
        Binding("ctrl+insert", "copy_to_clipboard"),
        Binding("ctrl+p", "print"),
    ]

    def compose(self) -> ComposeResult:
        yield Output()

    @work(exclusive=True, thread=True)
    def action_copy_to_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        print("Copy to clipboard...")

    @work(exclusive=True, thread=True)
    def action_print(self) -> None:
        self.query_one(Output).begin_capture_print()

        print("Ctrl + p...")

class Output(RichLog):
    def __init__(self) -> None:
        super().__init__(highlight=True, markup=True)

    @on(events.Print)
    def on_print(self, event: events.Print) -> None:
        if event.text.strip():
            self.write(event.text)


app = ScriptApp()

if __name__ == "__main__":
    app.run()
```

Run this app, the ctrl + insert binding doesn't work but ctrl + p works.

**Comment by davep** [comment]
Please see [this FAQ](https://textual.textualize.io/FAQ/#why-do-some-key-combinations-never-make-it-to-my-app).

**Reply by Tim-Cao** [reply]
Hi @davep, the shift+insert works in textual app. Also, the ctrl+insert and shift+insert works in my terminal application `Teminater`
Would you mind double checking?

**Reply by davep** [reply]
Double-checking what exactly?

**Reply by Tim-Cao** [reply]
Checking these
- the `ctrl+insert` could copy selected text in Terminator application, 
- the `shift+insert` could paste from clipboard in Terminator application. 
- But `ctrl+insert` cannot copy selected text in textual app.

**Reply by davep** [reply]
Your question is about creating bindings within Textual applications, right? Did you read the FAQ entry I pointed you at?

**Reply by Tim-Cao** [reply]
Sorry for confusion. My question is not creating bindings.
I met the default shortcuts works in Terminal applications like Terminator but not textual app.
- the ctrl+insert could copy selected text in Terminator application,
- the shift+insert could paste from clipboard in Terminator application.
- But ctrl+insert cannot copy selected text in textual app.

Both functions are testing shortcuts work or not
```python
    @work(exclusive=True, thread=True)
    def action_copy_to_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        print("Copy to clipboard...")

    @work(exclusive=True, thread=True)
    def action_print(self) -> None:
        self.query_one(Output).begin_capture_print()

        print("Ctrl + p...")
```

**Reply by davep** [reply]
Your example code literally used a binding to attempt to run that action:

```python
Binding("ctrl+insert", "copy_to_clipboard"),
```

I'll ask again: did you read the FAQ entry, do you understand what it's saying; did you follow the final bit of advice it gave?

**Reply by Tim-Cao** [reply]
@davep Yes, I read it.
Let's use this one
```python
from textual import on, work
from textual.app import App, ComposeResult, events
from textual.widgets import *
from textual.binding import Binding


class ScriptApp(App):
    BINDINGS = [
        Binding("ctrl+insert", "copy_to_clipboard", show=False),
        Binding("shift+insert", "paste_from_clipboard"),
    ]

    def compose(self) -> ComposeResult:
        yield TextArea("Test Sceniaro 1:")
        yield Output()

    @work(exclusive=True, thread=True)
    def action_copy_to_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        print("Copy to clipboard...")


class Output(RichLog):
    def __init__(self) -> None:
        super().__init__(highlight=True, markup=True)

    @on(events.Print)
    def on_print(self, event: events.Print) -> None:
        if event.text.strip():
            self.write(event.text)


app = ScriptApp()

if __name__ == "__main__":
    app.run()
```

ctrl+insert doesn't work but shift+insert works
According to test key events, the insert key could trigger event
![image](https://github.com/Textualize/textual/assets/52661397/50e8f72d-0fe3-4e13-9b08-062aa43ab31f)

It's so weird, why ctrl+insert doesn't work?

Does this list mean the insert is out of support list?
![image](https://github.com/Textualize/textual/assets/52661397/06e0ec8a-5658-4518-a9c9-edccd1dacc69)

**Reply by Tim-Cao** [reply]
It seems cannot override shortcuts with insert key
```python
from textual import on, work
from textual.app import App, ComposeResult, events
from textual.widgets import *
from textual.binding import Binding


class ScriptApp(App):
    BINDINGS = [
        Binding("ctrl+insert", "copy_to_clipboard", show=False),
        Binding("shift+insert", "paste_from_clipboard"),
    ]

    def compose(self) -> ComposeResult:
        yield TextArea("Test Sceniaro 1:")
        yield Output()

    @work(exclusive=True, thread=True)
    def action_copy_to_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        print("Copy to clipboard...")

    @work(exclusive=True, thread=True)
    def action_paste_from_clipboard(self) -> None:
        self.query_one(Output).begin_capture_print()

        print("Paste to clipboard...")

class Output(RichLog):
    def __init__(self) -> None:
        super().__init__(highlight=True, markup=True)

    @on(events.Print)
    def on_print(self, event: events.Print) -> None:
        if event.text.strip():
            self.write(event.text)


app = ScriptApp()

if __name__ == "__main__":
    app.run()
```

**Reply by davep** [reply, marked as answer]
> Does this list mean the insert is out of support list?

In my experience the insert key, and modified forms of it, are not fully-supported across the multitude of terminal emulators that are available. For example, here is insert, shift+insert and ctrl+insert as recorded under rio:

![Screenshot 2024-01-20 at 15 20 16](https://github.com/Textualize/textual/assets/28237/362dc8d3-ba91-4b57-9445-26956cc29211)

This isn't because of anything Textual is doing; this is because of what rio sends. 

The advice given in the FAQ is there for the reason it states: not all key combinations will work under all terminal emulators, because of how those terminal emulators work (and sometimes because the keys are special to the host operating system).

**Reply by Tim-Cao** [reply]
I see. Thanks @davep
